### PR TITLE
Switch from ZMQ to Websockets

### PR DESCRIPTION
**Pull request by hugobho**
Update to ExoRelay version using Websockets, and remove old ZMQ code.

@kevgo @trushton @martinjaime @helenmedrano

### CODE REVIEW

File: features/example-services/with-before-all-hook/service.yml [hunk 1]
```diff
@@ -1,3 +1,3 @@
-name: before-all hook service
+name: with-before-all-hook
```
Comment (trushton) [inline features/example-services/with-before-all-hook/service.yml]: Do we know why these names had to be changed if they were working before?

File: features/configuring-the-port.feature [hunk 2]
```diff
@@ -7,15 +7,14 @@ Feature: Defining the port at which the server listens
...
 
 
   Background:
-    Given an ExoCom instance
+    Given an ExoCom instance running at port 3001
 
 
-  Scenario: Running at a custom port
+  Scenario: Running at a given ExoCom port
     When starting a service at port 3001
```
Comment (trushton) [inline features/configuring-the-port.feature]: This reads funny to me. The service isn't really starting at port 3001, its simply connecting to the Exocom server at port 3001. IMO it would make more sense for the `Given an ExoCom instance running at port 3001` on line 15 to specify the port, and then for this step to just say `When starting a service`. What do you think?

File: features/example-services/es6/service.yml [hunk 3]
```diff
@@ -1,3 +1,3 @@
-name: es6 service
+name: es6
```
Comment (trushton) [inline features/example-services/es6/service.yml]: And here

File: features/example-services/test/service.yml [hunk 4]
```diff
@@ -1,3 +1,3 @@
-name: test service
+name: test
```
Comment (trushton) [inline features/example-services/test/service.yml]: And here

File: features/example-services/with-before-all-hook/service.yml (hunk 1 above)
Comment (hugobho) [inline features/example-services/with-before-all-hook/service.yml]: These name were never actually used for anything important before. Now, Exoservice looks at the `service.yml` for the service name and sends this along to ExoCom, where it is used as the hash key to identify the actual service within ExoCom. So in the feature tests we were giving the service one name, and in the `service.yml` file a slightly different name so ExoCom thought they were different services.

File: features/configuring-the-port.feature (hunk 2 above)
Comment (hugobho) [inline features/configuring-the-port.feature]: So like:

```
Given an ExoCom instance running at port 3001
When starting a service
Then it comes online at port 3001
```
?

Comment (trushton) [inline features/configuring-the-port.feature]: I was thinking more along the lines of just:
```
Given an ExoCom instance running at port 3001
When starting a service
Then it comes online
```
As long as the `ExoCom instance running at port 3001` step passes, we already know that that is the same port that the service will be connecting to.

File: features/example-services/with-before-all-hook/service.yml (hunk 1 above)
Comment (trushton) [inline features/example-services/with-before-all-hook/service.yml]: 👍

File: features/configuring-the-port.feature (hunk 2 above)
Comment (hugobho) [inline features/configuring-the-port.feature]: Wouldn't it still be important to verify that the service comes online at port 3001? Since this is the `configuring-the-port.feature` file. Plus Exoservice can be executed independent of `exo-run` and `exo-deploy` so shouldn't we make sure that when users pass a port as an environment variable, that it'll actually come online at that port? In other words, even if ExoCom comes online at port 3001 there's no guarantee that the service will come online at the same port.

Comment (kevgo) [inline features/configuring-the-port.feature]: @trushton @hugobho please make sure the feature specs convey meaningful information to the reader and not just trivialities. Imagine you don't know how exoservice works and ask me what exactly the ports that you give it mean, and then all I tell you is "When starting a service, then it comes online". This isn't a helpful answer because you kind of already figured that much. 

What do you think about this:

```cucumber
Background:
  Given an ExoCom instance running at port 3001

Scenario: an ExoService instance connects to ExoCom
  When starting a service at port 3001
  Then it connects to the ExoCom instance

Scenario: a wrong Exocom port is given
  When starting a service at port 3002
  Then it aborts with the error message "Cannot find ExoCom at port 3002"
  And it does not connect to the ExoCom instance
```

File: features/example-services/es6/service.yml (hunk 3 above)
Comment (kevgo) [inline features/example-services/es6/service.yml]: Why do you rename this service?

File: features/example-services/with-before-all-hook/service.yml (hunk 1 above)
Comment (kevgo) [inline features/example-services/with-before-all-hook/service.yml]: Thanks for unifying the names. Let's keep them meaningful, however. The service names will be used in many places in the future: the "service store" where one can browse for services to add to an application etc. It is therefore important that the service names are meaningful, i.e. say "user service" instead of just "user". The feature specs here give examples for everything that other developers use as a reference, so it is important to get these names right here. 

What do you think about "example service with before-all hook"?

File: features/steps/steps.ls [hunk 5]
```diff
@@ -21,11 +21,15 @@ module.exports = ->
         done!
 
 
+  @Given /^an ExoCom instance running at port (\d+)$/, (@exocom-port, done) ->
+    @exocom = new ExoComMock
+      ..listen @exocom-port
+      done!
```
Comment (kevgo) [inline features/steps/steps.ls]: This line has the wrong indentation

File: features/example-services/es6/service.yml (hunk 3 above)
Comment (trushton) [inline features/example-services/es6/service.yml]: Related to your first comment, I was unsure as to why this was renamed from es6 service to just es6, but that has been cleared up now.

**Comment by trushton** [comment]
@kevgo, @hugobho is handing this PR over to me so that he can finish up work on `exo-deploy`. I have made the changes as requested above, except for the one regarding the name in service.yml as we still need to figure out the name vs type convention for services in the future. This version should now be compatible with the latest updates to `exorelay` and `exocom-dev`.

File: features/before-all-hook.feature [hunk 6]
```diff
@@ -13,5 +13,5 @@ Feature: Before-All hook
 
 
   Scenario: code with beforeAll hook
-    When starting the "with-before-all-hook" service
+    When starting the "example service with before-all hook"
```
Comment (kevgo) [inline features/before-all-hook.feature]: Please add the word "service" at the end of this step, I think it makes it more readable

File: features/configuring-the-port.feature [hunk 7]
```diff
@@ -7,15 +7,17 @@ Feature: Defining the port at which the server listens
...
 
 
   Background:
-    Given an ExoCom instance
+    Given an ExoCom instance running at port 3001
 
-
-  Scenario: Running at a custom port
+  Scenario: an ExoService instance connects to ExoCom
     When starting a service at port 3001
```
Comment (kevgo) [inline features/configuring-the-port.feature]: When starting a service configured for ExoCom port 3001

File: features/example-services/es6/server.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
 module.exports = {
 
   ping: (_, {reply}) => {
-    reply('pong')
+    reply('pong', '')
```
Comment (kevgo) [inline features/example-services/es6/server.js]: Isn't the second parameter optional?

File: features/example-services/es6/service.yml (hunk 3 above)
Comment (kevgo) [inline features/example-services/es6/service.yml]: Please undo the change on this line

File: features/example-services/test/service.yml (hunk 4 above)
Comment (kevgo) [inline features/example-services/test/service.yml]: Please undo the changes on this line

File: features/example-services/test/src/server.ls [hunk 9]
```diff
@@ -6,7 +6,7 @@ module.exports =
 
 
   ping: (_, {reply}) ->
-    reply 'pong'
+    reply 'pong', ''
```
Comment (kevgo) [inline features/example-services/test/src/server.ls]: The second parameter should be optional

File: features/steps/steps.ls [hunk 10]
```diff
@@ -21,11 +21,15 @@ module.exports = ->
         done!
 
 
+  @Given /^an ExoCom instance running at port (\d+)$/, (@exocom-port, done) ->
+    @exocom = new ExoComMock
+      ..listen @exocom-port
+    done!
```
Comment (kevgo) [inline features/steps/steps.ls]: This step is not asynchronous, remove `done` from here and the parameter.

File: features/steps/steps.ls [hunk 11]
```diff
@@ -54,20 +58,18 @@ module.exports = ->
...
+    @create-exoservice-instance {@service-name, @exocom-port}, ~>
+      @remove-register-service-message @exocom, done
 
 
-  @When /^starting a service at port (\d+)$/, (exorelay-port, done) ->
+  @When /^starting a service at port (\d+)$/, (port, done) ->
     @service-name = 'test'
-    @exocom.register-service {name: @service-name, port: exorelay-port}
-    @create-exoservice-instance {@service-name, exorelay-port, @exocom-port}, ~>
+    @create-exoservice-instance {@service-name, @exocom-port}, ~>
```
Comment (kevgo) [inline features/steps/steps.ls]: Shouldn't this say `exocom-port: port` instead of `@exocom-port` here?

File: features/steps/steps.ls [hunk 12]
```diff
@@ -106,8 +107,11 @@ module.exports = ->
...
 
-  @Then /^the service runs at port (\d+)$/, (port, done) ->
+  @Then /^it connects to the ExoCom instance$/, (done) ->
     @exocom.send service: 'test', name: '__status' , id: '123'
     wait-until (~> @exocom.received-messages[0]), 1, ~>
       if @exocom.received-messages[0].name is "__status-ok"
         done!
+
+  @Then /^it aborts with the error message "([^"]*)"$/ (error-message, done) ->
+    process.wait error-message, done!
```
Comment (kevgo) [inline features/steps/steps.ls]: Remove the `!` after `done`, otherwise it doesn't wait

File: features/example-services/es6/server.js (hunk 8 above)
Comment (trushton) [inline features/example-services/es6/server.js]: It used to be, but in `exorelay`'s `reply-method-for` method, we changed the lines 
```
(message-name, payload) ~>
  @send message-name, payload, response-to: id
```
to `@send _, _, response-to: id` and if no second parameter is given, this doesn't perform as desired.

Should we revert that change in `exorelay`?

Comment (kevgo) [inline features/example-services/es6/server.js]: No, please add a default value of `null` for the `payload` parameter to something non-null:

```ls
(message-name, payload = {}) ~>
  @send _, _, response-to: id

**Comment by trushton** [comment]
@kevgo All comments addressed

File: features/configuring-the-port.feature [hunk 13]
```diff
@@ -7,15 +7,17 @@ Feature: Defining the port at which the server listens
 
   Rules:
   - call "exo-js run" in the directory of an Exosphere service
-    to run the service at the default port 3000
-  - if port 3000 is taken, it chooses the next available port above that
-  - the port can be customized via the "--port" command-line switch
+    to connect the service to the ExoCom host and port
```
Comment (kevgo) [inline features/configuring-the-port.feature]: This is no longer correct. One has to provide the ExoCom host and port via environment variables now.

```
- call "exo-js run" in the directory of an Exosphere service to start up that service
- provide the ExoCom address via the environment variables EXOCOM_HOST and EXOCOM_PORT

File: features/steps/steps.ls [hunk 14]
```diff
@@ -14,18 +14,21 @@ require! {
 
 module.exports = ->
 
-  @Given /^an ExoCom instance$/, (done) ->
+  @Given /^an ExoCom instance$/, ->
```
Comment (kevgo) [inline features/steps/steps.ls]: This is an async method. Please add a `done` parameter and call it at the end.

File: features/steps/steps.ls [hunk 15]
```diff
@@ -14,18 +14,21 @@ require! {
...
     port-reservation.get-port N (@exocom-port) ~>
       @exocom = new ExoComMock
         ..listen @exocom-port
-        done!
+
+
+  @Given /^an ExoCom instance running at port (\d+)$/, (@exocom-port, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: This is a synchronous method. Please remove the `done` argument.

File: features/steps/steps.ls [hunk 16]
```diff
@@ -53,24 +56,28 @@ module.exports = ->
...
+      @remove-register-service-message @exocom, done
 
 
-  @When /^starting a service at port (\d+)$/, (exorelay-port, done) ->
-    @service-name = 'test'
-    @exocom.register-service {name: @service-name, port: exorelay-port}
-    @create-exoservice-instance {@service-name, exorelay-port, @exocom-port}, ~>
+  @When /^starting a service configured for ExoCom port (\d+)$/, (port, done) ->
+    @service-name = 'test service'
+    console.log @service-name
```
Comment (kevgo) [inline features/steps/steps.ls]: Remove this line

File: features/steps/steps.ls [hunk 17]
```diff
@@ -106,8 +112,12 @@ module.exports = ->
...
-  @Then /^the service runs at port (\d+)$/, (port, done) ->
-    @exocom.send service: 'test', name: '__status' , id: '123'
+  @Then /^it connects to the ExoCom instance$/, (done) ->
+    @exocom.send service: @service-name, name: '__status' , id: '123'
     wait-until (~> @exocom.received-messages[0]), 1, ~>
       if @exocom.received-messages[0].name is "__status-ok"
         done!
+
+  @Then /^it aborts with the error message "([^"]*)"$/ (error-message, done) ->
+    expect(console.log error-message)
```
Comment (kevgo) [inline features/steps/steps.ls]: This step is not doing what it is supposed to.
For the sake of getting this shipped, let's add a `# TODO: implement this` tag here. This isn't a super critical test.

**Comment by trushton** [comment]
Last round comments addressed.
